library/std/src/sys/dbg.rs

```diff
@@ -0,0 +1,186 @@
+            Some(DebuggerPresence::Detected)
+        } else {
+            Some(DebuggerPresence::NotDetected)
+        }
+    }
+}
+
+#[cfg(target_os = "linux")]
+mod os {
+    use super::DebuggerPresence;
+    use crate::fs::File;
+    use crate::io::{BufRead, BufReader};
+
+    pub(super) fn is_debugger_present() -> Option<DebuggerPresence> {
+        let file = File::open("/proc/self/status").ok()?;
+        let reader = BufReader::new(file);
+
+        for line in reader.lines().flatten() {
+            if line.starts_with("TracerPid:") {
+                return if line.split(':').nth(1).map_or(false, |pid| pid.trim() != "0") {
+                    Some(DebuggerPresence::Detected)
+                } else {
+                    Some(DebuggerPresence::NotDetected)
+                };
+            }
+        }
+
+        None
+    }
+}
+
+// 32-bit macOS Rust targets are no longer shipped.
+#[cfg(not(any(
+    target_os = "windows",
+    all(target_os = "macos", target_pointer_width = "64"),
+    target_os = "freebsd",
+    target_os = "linux"
+)))]
+mod os {
+    pub(super) fn is_debugger_present() -> Option<super::DebuggerPresence> {
+        None
+    }
+}
+
+/// Detect the debugger presence.
+///
+/// The code does not try to detect the debugger at all costs (e.g., when anti-debugger
+/// tricks are at play), it relies on the interfaces provided by the OS.
+///
+/// Return value:
+/// * `None`: it's not possible to conclude whether the debugger is attached to this
+///    process or not. When checking for the presence of the debugger, the detection logic
+///    encountered an issue, such as the OS API throwing an error or the feature not being
+///    implemented.
+/// * `Some(DebuggerPresence::Detected)`: yes, the debugger is attached
+///   to this process.
+/// * `Some(DebuggerPresence::NotDetected)`: no, the debugger is not
+///    attached to this process.
+pub(crate) fn is_debugger_present() -> Option<DebuggerPresence> {
+    if cfg!(miri) { None } else { os::is_debugger_present() }
+}
+
+/// Execute the breakpoint instruction if the debugger presence is detected.
+/// Useful for breaking into the debugger without the need to set a breakpoint
+/// in the debugger.
+///
+/// Note that there is a race between attaching or detaching the debugger, and running the
+/// breakpoint instruction. In case the debugger is attached and the function is about
+/// to run the breakpoint instruction yet right before that the debugger detaches, the
+/// process will crash due to running the breakpoint instruction and the debugger not
+/// handling the trap exception.
+pub(crate) fn breakpoint_if_debugging() -> Option<DebuggerPresence> {
+    let debugger_present = is_debugger_present();
+    if let Some(DebuggerPresence::Detected) = debugger_present {
+        // SAFETY: Executing the breakpoint instruction. No state is shared
+        // or modified by this code.
+        unsafe {
+            core::intrinsics::breakpoint();
+        }
```

Comment:
pure nit, but:
```suggestion
        unsafe { core::intrinsics::breakpoint() };
```

Reply:
I'll try to do that (my guess is that `cargo fmt` should've formatted the line on saving) and report back

Reply:
Couldn't make that stick with `cargo fmt` running on save. VSCode and helix format the file differently as a surprise to me. Running `cargo fmt` from the repo's root dir changes some 200 files. In the end, used `vim` to change that file, and committed the result.